base/operators.jl

```diff
@@ -1019,12 +1019,20 @@ end
 
 function (c::ComposedFunction)(x...; kw...)
     fs = unwrap_composed(c)
-    call_composed(fs[1](x...; kw...), tail(fs)...)
+    fs[1](call_composed(tail(fs), x, kw))
 end
-unwrap_composed(c::ComposedFunction) = (unwrap_composed(c.inner)..., unwrap_composed(c.outer)...)
+eval(quote
+    # @assume_effects :terminates_globally, but @assume_effects cannot be used
+    # Core.Compiler, hence the Expr(:meta, ...)
+    function unwrap_composed(c::ComposedFunction)
+        #$(Expr(:meta, :($(Expr(:purity, false, false, false, true, false, false)))))
+        $(Expr(:meta, Expr(:purity, false, false, false, true, false, false)))
+        return (unwrap_composed(c.outer)..., unwrap_composed(c.inner)...)
+    end
+end)
```

Comment:
I think we generally want to avoid direct creation of `:purity` expression in this way as it is fragile against future changes in the effect system design.

I'd like to add this diff:
```diff
diff --git a/base/essentials.jl b/base/essentials.jl
index 906c36ad9c..8eee2923b3 100644
--- a/base/essentials.jl
+++ b/base/essentials.jl
@@ -204,23 +204,35 @@ macro _pure_meta()
 end
 # can be used in place of `@assume_effects :total` (supposed to be used for bootstrapping)
 macro _total_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#true,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#true))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        nothrow=true,
+        terminates_globally=true,
+        notaskstate=true))
 end
 # can be used in place of `@assume_effects :foldable` (supposed to be used for bootstrapping)
 macro _foldable_meta()
-    return _is_internal(__module__) && Expr(:meta, Expr(:purity,
-        #=:consistent=#true,
-        #=:effect_free=#true,
-        #=:nothrow=#false,
-        #=:terminates_globally=#true,
-        #=:terminates_locally=#false,
-        #=:notaskstate=#false))
+    return _is_internal(__module__) && Expr(:meta, _assumed_effects(;
+        consistent=true,
+        effect_free=true,
+        terminates_globally=true))
+end
+
+function _assumed_effects(;
+    consistent::Bool = false,
+    effects_free::Bool = false,
+    nothrow::Bool = false,
+    terminates_globally::Bool = false,
+    terminates_locally::Bool = false,
+    notaskstate::Bool = false)
+    return Expr(:purity,
+        consistent,
+        effects_free,
+        nothrow,
+        terminates_globally,
+        terminates_locally,
+        notaskstate)
 end
 
 # another version of inlining that propagates an inbounds context
```

and then rewrite this piece of code like:
```suggestion
eval(quote
    # equivalent to `@assume_effects :terminates_globally`,
    # but @assume_effects cannot be used `Core.Compiler`, hence `_assumed_effects(...)`
    function unwrap_composed(c::ComposedFunction)
        $(Expr(:meta, _assumed_effects(; terminates_globally=true)))
        return (unwrap_composed(c.outer)..., unwrap_composed(c.inner)...)
    end
end)
```

Then we don't look for all the usages of `:purity` expression when we add new effect to the system as far as we keep the same naming schema for the existing effects.

Reply:
It seems more difficult than I expected to use kwfunc until we define namedtuple.jl.. We can leave this to the future and don't need to let it hold this PR.

Reply:
Alternatively, we could move the `ComposedFunction` stuff to a dedicated file that is included after `array.jl` (which makes `@assume_effects` work) but before `bitarray.jl` (which uses composition with `!iszero`).